src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Package/Update/VersionChooser.cs

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Common;
+using NuGet.Configuration;
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+using NuGet.Versioning;
+
+namespace NuGet.CommandLine.XPlat.Commands.Package.Update
+{
+    internal static class VersionChooser
+    {
+        internal static async Task<NuGetVersion?> GetLatestVersionAsync(
+            string packageId,
+            CachingSourceProvider sourceProvider,
+            ISettings settings,
+            SourceCacheContext sourceCacheContext,
+            ILoggerWithColor logger,
+            CancellationToken cancellationToken)
+        {
+            var sources = new List<SourceRepository>();
+            foreach (PackageSource packageSource in SettingsUtility.GetEnabledSources(settings).NoAllocEnumerate())
+            {
+                SourceRepository sourceRepository = sourceProvider.CreateRepository(packageSource);
+                sources.Add(sourceRepository);
+            }
+
+            var lookups = new Task<NuGetVersion?>[sources.Count];
+            for (int source = 0; source < sources.Count; source++)
+            {
+                SourceRepository sourceRepository = sources[source];
+                // If package source is a local folder feed, it might not actually be async
+                lookups[source] = Task.Run(() => FindHighestPackageVersionAsync(sourceRepository, packageId, sourceCacheContext, logger, cancellationToken));
+            }
+
+            await Task.WhenAll(lookups);
+
+            NuGetVersion? highestVersion = null;
+            foreach (var task in lookups)
+            {
+                if (task.Result != null)
+                {
+                    if (highestVersion == null || task.Result > highestVersion)
+                    {
+                        highestVersion = task.Result;
+                    }
+                }
+            }
+
+            return highestVersion;
+        }
+
+        private static async Task<NuGetVersion?> FindHighestPackageVersionAsync(
+            SourceRepository source,
+            string packageId,
+            SourceCacheContext sourceCacheContext,
+            ILogger logger,
+            CancellationToken cancellationToken)
+        {
+            var packageMetadataResource = await source.GetResourceAsync<PackageMetadataResource>(cancellationToken);
+
+            var packageDetails = await packageMetadataResource.GetMetadataAsync(packageId, false, false, sourceCacheContext, logger, cancellationToken);
+
+            if (packageDetails is null || !packageDetails.Any())
+            {
+                return null;
+            }
+
+            NuGetVersion highestVersion = packageDetails.Max(p => p.Identity.Version)!;
```

Comment:
This highest version could be that is not compatible with the project framework, we could look at the `DependencySets` and see if this package is compatible with the tfm, if not we could try until we find it. 

This is going to be more expensive but would improve the success rate of restore, this is one of the problems we have with the PM UI where we don't validate tfm compatibility.

Reply:
As mentioned in the PR description, this initial version was intentionally minimal to try to meet preview deadlines. A tracking issue was already created, which also discusses that dependency groups is not the same thing as package compatibility and in some scenarios will give the wrong results:
* https://github.com/NuGet/Home/issues/14312